File: python/e3.py
"""
Finds the largest prime factor of the number 600851475143.
"""
import numpy as np
from primegen import primes_by_sieve

def Euler_3(n = 600851475143):
    #generate all the candidate primes
    max_num = int(np.sqrt(n))
    primes = primes_by_sieve(max_num)
    return find_biggest_factor_functional(primes, n)

def find_biggest_factor_reversed(primes, n):
    while True:
        for prime in primes[::-1]:
            if n % prime == 0:
                return prime

def find_biggest_factor_functional(primes, n):
    return max(p for p in primes if n % p == 0)

def test_3():
    assert Euler_3() == 6857

if __name__ == "__main__":
    print(Euler_3())

File: rust/e3.rs
/*
Finds the largest prime factor of the number 600851475143.
*/
mod primegen;

fn euler_3(big_num : usize) -> usize {
    //find largest prime factor
    let upper_limit = (big_num as f64).sqrt().ceil() as usize;
    let primes = primegen::primes_by_sieve(upper_limit);
    *primes.iter().filter(|&p| big_num % p == 0).max().unwrap()
}

fn main() {
    println!("{}", euler_3(600_851_475_143));
}